Edge: fetch name of aria-alert text by traversing its children, which then allows live region changed event to work.

josephsl opened the pull request:
### Link to issue number:
Resolves #8466 

### Summary of the issue:
In Edge, aria-role=alert isn't announcement properly due to the fact that there is no name for this element.

### Description of how this pull request fixes the issue:
When alerts are received, Edge fires live region changed event. However, the name of this element is empty, but thankfully, its children's names are text fragments. Therefore set name property by traversing its children and combine them, which then allows live region changed event to work.

### Testing performed:
Tested on Edge in Windows 10 Version 1803 and 1709.

### Known issues with pull request:
None

### Change log entry:
Bug fixes: In Microsoft Edge, NVDA will announce alerts marked with aria-role=alert property.

josephsl committed:
Edge: fetch name of aria-alert text by traversing its children, which then allows live region changed event to work. Re #8466.

When alerts are received, Edge fires live region changed event. However, the name of this element is empty, but thankfully, its children's names are text fragmenets. Therefore set name property by traversing its children, which then allows live region changed event to work.

LeonarddeR reviewed (changes requested):
Your current proposal creates a new python NVDAObject for every child and only deals with children at one level deep. I propose at least one of the following:

1. Create a custom cache request to retrieve the children, similar to how this is done in the outlook app module.
2. We're deealing with edge node objects, which have a tree interceptor. You can also retrieve the text by using obj.treeInterceptor.makeTextInfo(obj).text

josephsl committed:
EdgeNode/aria-role=alert: address review action. Re #8466.

Reviewed by @LeonardDer (Babbage): because concatenation only works on first-level descendant, it may not pick up additional text from deeper descendants. Therefore use tree interceptor content as the element's name, which still allows live region changed event to work (thank you Leonard for this tip).

leonardder commented on source/NVDAObjects/UIA/edge.py:
```diff
@@ -422,6 +422,13 @@ def getNormalizedUIATextRangeFromElement(self,UIAElement):
+	def _get_name(self):
+		name=super(EdgeNode,self).name
+		# #8466: elements with aria-role=alert set fires live region changed event but does not expose the alert text as its name, wihch can be found in its descendants.
+		if "role" in self.ariaProperties and self.ariaProperties["role"] == "alert":
```
I think it might be good to check for the tree interceptor's existence, just to make sure. How about:

`if self.ariaProperties.get("role") == "alert" and self.treeInterceptor:`

josephsl commented:
Hi, ah yes, I thought about that for a second before making that change… Point taken, thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Monday, July 2, 2018 11:21 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Edge: fetch name of aria-alert text by traversing its children, which then allows live region changed event to work. (#8472)

 

@leonardder commented on this pull request.

  _____  

In source/NVDAObjects/UIA/edge.py <https://github.com/nvaccess/nvda/pull/8472#discussion_r199581673> :

> @@ -422,6 +422,13 @@ def getNormalizedUIATextRangeFromElement(self,UIAElement):
                       charInfo.collapse(True)
               return range
 
+       def _get_name(self):
+              name=super(EdgeNode,self).name
+              # #8466: elements with aria-role=alert set fires live region changed event but does not expose the alert text as its name, wihch can be found in its descendants.
+              if "role" in self.ariaProperties and self.ariaProperties["role"] == "alert":

I think it might be good to check for the tree interceptor's existence, just to make sure. How about:

if self.ariaProperties.get("role") == "alert" and self.treeInterceptor:

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8472#pullrequestreview-133729372> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkLphJQCRT1qbl1kNPog_WDPHjyAEks5uCmR_gaJpZM4U-L7m> .

josephsl committed:
EdgeNode/aria-alert: make sure tree interceptor is alive. Re #8466.

Reviewed by @LeonardDer (Babbage): just to make sure: tree interceptor must be live in order for tree interceptor based solution to work.

LeonarddeR reviewed (dismissed):
Note that review from @michaeldcurran is still required.

josephsl committed:
Merge branch 'master' into i8466edgeAriaAlert

josephsl committed:
Merge branch 'master' into i8466edgeAriaAlert

josephsl committed:
What's new/bug fixes: add Edge/aria-role=alert entry. Re #8466.

josephsl committed:
Merged master

josephsl commented:
Hi,

Besides resolving conflicts, I'll update what's new to include #8557, as that one is also identical to this one except that it deals with system alert event.

Thanks.

josephsl committed:
What's new: add issue #8557 (system alert event), as the underlying cause is the same.

josephsl commented:
Hi,

To @LeonardDer: requesting another round of review please. Thanks.

michaelDCurran reviewed (changes requested):
I'm not comfortable with this approach. I think I would rather see code specific to liveRegionchange and SystemAlert events on this object, rather than overriding the name property. Reasons being:
1. The object could already have an ARIA label or other author-provided name, which now gets lost.
2. There could be a lot of text fetched in that object. In the past we have had problems with browsers that calculate name from descendants too much.

josephsl committed:
EdgeNode/live region and system alert events: address review actions.

Requested by Mick Curran (NV Access): what if authors provide cusotm ARIA labels for the element in question? As of now, the name property does not take this into account. Instead, handlers for UIA system laert and live region change events will be defined with room for expansion when the need arises.

josephsl commented:
Hi,

Right. For now, I'll define handlers for both of these, with NVDA announcing tree interceptor content (unless it has performance issues). This is so that improvements can be made as we encounter more alerts in the wild.

Thanks for catching this flaw.

leonardder commented:
In which cases do UIA_systemAlert or liveRegionChanged events apply to these alerts? Couldn't this lead to duplicated announcements somehow?

josephsl commented:
Hi, if I got this correctly, #8466’s test case is a classic case of live regions, while #8557 uses system alert event. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Wednesday, July 25, 2018 11:06 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Edge: fetch name of aria-alert text by traversing its children, which then allows live region changed event to work. (#8472)

 

In which cases do UIA_systemAlert or liveRegionChanged events apply to these alerts? Couldn't this lead to duplicated announcements somehow?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8472#issuecomment-407986657> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkEWHYTQHZkjdb8bCb0o13Y2WUDteks5uKVw0gaJpZM4U-L7m> .

josephsl committed:
Merged master

leonardder commented on source/NVDAObjects/UIA/edge.py:
```diff
@@ -516,6 +517,16 @@ def _get_landmark(self):
 
+	def event_liveRegionChange(self):
+		# #8466: elements with aria-role=alert set fires live region changed event but does not expose the alert text as its name, wihch can be found in its descendants.
+		if self.ariaProperties.get("role") == "alert" and self.treeInterceptor:
```
Wouldn't this make live region events go silent if this predicate is not True? I'd say you have to check that the object has no name as well. furthermore, if the predicate is False, please call the super method.

leonardder commented on source/NVDAObjects/UIA/edge.py:
```diff
@@ -516,6 +517,16 @@ def _get_landmark(self):
+	def event_UIA_systemAlert(self):
+		# #8557: similar to live region change event, some elements does not have a name but text is scattered among its descendants.
+		if self.ariaProperties.get("role") == "alert" and self.treeInterceptor:
+			ui.message(self.treeInterceptor.makeTextInfo(self).text)
```
Same concerns apply here.

leonardder commented on source/NVDAObjects/UIA/edge.py:
```diff
@@ -516,6 +517,16 @@ def _get_landmark(self):
 		return None
 
+	def event_liveRegionChange(self):
+		# #8466: elements with aria-role=alert set fires live region changed event but does not expose the alert text as its name, wihch can be found in its descendants.
```
Grammar: does > do

josephsl committed:
EdgeNode: call superclass event handler for live region change and system alert.

Caught by Leonard de Juijter (Babbage): live region change and system alert will fail if the alert is an aria alert, there is a name, and there is no tree interceptor. Thus catch this and call the superclass method.

josephsl committed:
Merged master

josephsl committed:
EdgeNode: comment grammar edits

josephsl committed:
Merged master

josephsl committed:
What's new: 2018.4 section.

josephsl committed:
Merged master

josephsl committed:
What's new: remove bug fix entry for Edge+aria alert.

leonardder commented:
@josephsl Plans to continue this pr? Probably not of very high importance due to current Edge being abandoned.

josephsl commented:
Hi, probably unlikely unless progressive web apps require this PR. So let’s keep this aside for now. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Wednesday, November 27, 2019 12:41 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Edge: fetch name of aria-alert text by traversing its children, which then allows live region changed event to work. (#8472)

 

@josephsl <https://github.com/josephsl>  Plans to continue this pr? Probably not of very high importance due to current Edge being abandoned.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8472?email_source=notifications&email_token=AB4AXEBBZQIAVU4PMHBOLHDQV3LNPA5CNFSM4FHYX3TKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFKWF7Q#issuecomment-559244030> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEA7PN5MESCOHJD2UJ3QV3LNPANCNFSM4FHYX3TA> .

Adriani90 commented:
@feerrenrut maybe this is also worthy to mark as edge spartan for future reference, I cannot add labels to PRs.
